[clang] Merge gtest binaries into AllClangUnitTests

rnk opened the pull request:
This reduces the size of the clang/unittests build directory by 64% and my overall build dir size by 5%. Static linking is the real driving factor here, but even if the default build configuration used shared libraries, I don't see why we should be building so many unit test binaries.

To make the project more approachable for new contributors, I'm attempting to make the build a bit less resource-intensive. Build directory size is a common complaint, and this is low-hanging fruit.

rnk committed:
[clang] Merge gtest binaries into AllClangUnitTests

This reduces the size of the clang/unittests build directory by 64% and
my overall build dir size by 5%. Static linking is the real driving
factor here, but even if the default build configuration used shared
libraries, I don't see why we should be building so many unit test
binaries.

To make the project more approachable for new contributors, I'm
attempting to make the build a bit less resource-intensive. Build
directory size is a common complaint, and this is low-hanging fruit.

rnk committed:
Migrate new ParseTests unittest binary to new style

petrhosek commented on clang/unittests/CMakeLists.txt:
```diff
@@ -77,3 +104,25 @@ add_subdirectory(Index)
+)
+
+# The Tooling library has some internal headers. Make those work. If we like
+# the merged clang unit test binary, we can udpate the include paths and make
```
```suggestion
# the merged clang unit test binary, we can update the include paths and make
```

rnk committed:
Merge branch 'main' into merge-clang-unittests

rnk committed:
Interpret an empty code model as no code model, or the default code model

rnk committed:
Update clang/unittests/CMakeLists.txt

Co-authored-by: Petr Hosek <phosek@google.com>

rnk committed:
Initialize MCs so target machine construction works when targets are present

rnk committed:
Fix include lint

rnk committed:
Satisfy non-optional brief doc requirement for premerge

rnk committed:
Also set FULL_DOCS =S

rnk commented:
Ping, WDYT?

rnk committed:
Merge branch 'main' into merge-clang-unittests

rnk commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -304,7 +304,7 @@ getCodeModel(const CodeGenOptions &CodeGenOpts) {
                            .Case("medium", llvm::CodeModel::Medium)
                            .Case("large", llvm::CodeModel::Large)
-                           .Case("default", ~1u)
+                           .Cases("default", "", ~1u)
```
Basically, after my change, some Frontend tests now exercise this codepath without setting up a code model, and they hit the assert below. What changed is that now the target they are generating IR for is registered, so this codepath executes, when it didn't before my change. Interpreting "empty" as the default code model here seemed like the most sensible thing to do, rather than adding logic to those tests to explicitly set a code model.

rnk replied:
I responded on Discord, but just to resolve the comment here on github:
I recall there was a case where the main LLVM target gets registered, but the MCTargetDesc isn't registered, and this resulted in some failure along some codepath. Of course, if you can remove it and tests pass, anyone is welcome to revisit that. It's not obvious why this is needed.

rnk committed:
Merge branch 'main' into merge-clang-unittests

rnk committed:
Add null check to work around issue when the TargetInfo is registered, but not the TargetMachine

In this case, createTargetMachine may return null, which can happen in
some unit tests.

rnk commented:
If anyone is looking for stories about the value of premerge testing, I was all set to push this tonight, but I saw the linux premerge test failure. I debugged it, and it turns out to be a subtle interaction where if you run Clang codegen with registered targets but no registered TargetMachine classes, then one of our unittests will crash setting the large code model threshold 🙄 . Anyway, I pushed a fix and can land it tomorrow. Thanks for reviewing!

dyung commented:
> LLVM Buildbot has detected a new failure on builder `llvm-clang-aarch64-darwin` running on `doug-worker-4` while building `clang` at step 6 "test-build-unified-tree-check-all".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/190/builds/19202
> 
> Here is the relevant piece of the build log for the reference

@rnk not sure if you were notified or not, but the test you changed seems to be failing on MacOS. It seems to be consistently failing on my AArch64 MacOS bot, but only once randomly on my x86-64 MacOS bot (https://lab.llvm.org/buildbot/#/builders/23/builds/9847). Can you take a look?

rnk commented:
I attempted to repro on my macbook, but it doesn't reproduce. This test doesn't seem high value, and I thin kthe right next step is to disable it to green the bots before debugging it.

nico commented:
`BasicTests` and `FormatTests` had fairly few deps (they didn't need Sema etc) and compiled much fewer files than they presumably now do. Does it make sense to keep those as distinct binaries?

rnk commented:
> `BasicTests` and `FormatTests` had fairly few deps (they didn't need Sema etc) and compiled much fewer files than they presumably now do. Does it make sense to keep those as distinct binaries?

FormatTests has many clang dependencies, but after rebuilding, it looks like it's still quite small, so I sent a PR with my file size data:
https://github.com/llvm/llvm-project/pull/138021

You can see that the Interpreter tests are a much bigger target for optimization.

owenca commented:
This patch caused an assertion failure for b8bb1ccb4f9126d1bc9817be24e17f186a75a08b on several build bots and my mac (M2, 15.4.1):
```
[ RUN      ] CodeGenTest.TestNonAlterTest
Assertion failed: (MRI && "Unable to create reg info"), function initAsmInfo, file CodeGenTargetMachineImpl.cpp, line 48.
```
It also made us lose the `FormatTests` build target:
```
$ ninja FormatTests
ninja: error: unknown target 'FormatTests'
```

rnk commented:
Are the `CodeGenTest.TestNonAlterTest` failures still an issue? I don't see that failure on any bots. I haven't gotten any reports.

owenca commented:
@rnk see https://github.com/llvm/llvm-project/pull/137577#issuecomment-2843963228 for some of the bots that got the assertion failure. 

> Update: Back to normal after [afd738c](https://github.com/llvm/llvm-project/commit/afd738cd9016ddef4e26b309770ba35fd017a178).

I meant the assertion failure also disappeared after that commit.

AaronBallman commented:
> @rnk see [#137577 (comment)](https://github.com/llvm/llvm-project/pull/137577#issuecomment-2843963228) for some of the bots that got the assertion failure.
> 
> > Update: Back to normal after [afd738c](https://github.com/llvm/llvm-project/commit/afd738cd9016ddef4e26b309770ba35fd017a178).
> 
> I meant the assertion failure also disappeared after that commit.

FWIW, I'm seeing the test failure on Windows x64 with a debug build of Clang's unit tests as of a fresh fetch of `main` this morning.

dyung commented:
I've been seeing intermittent failures of AllClangUnitTests/TimeProfilerTest/ConstantEvaluationCxx20 and AllClangUnitTests/TimeProfilerTest/ConstantEvaluationC99 on my MacOS bot, but I haven't had a chance to try and figure out if it might be related to your change or not.

ConstantEvaluationCxx20:
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19393
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19381
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19356
ConstantEvaluationC99:
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19383
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19376
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19366
 - https://lab.llvm.org/buildbot/#/builders/190/builds/19343

First instance of either failing: https://lab.llvm.org/buildbot/#/builders/190/builds/19216

aeubanks commented:
I've also seen ConstantEvaluationC99 intermittently fail on our mac bots recently

AaronBallman commented:
Yup, seeing this again on https://github.com/llvm/llvm-project/pull/132776

rnk commented:
I hacked together a workaround here: https://github.com/llvm/llvm-project/pull/138613

I'm mostly just guessing remotely, since I don't have a reliable repro, that the final PerformPending... event is rounding into the previous event group that just ended. If the JSON encodes this nesting relationship already, it might be nice to expose that to deflake the test more.

tahonermann commented:
I ended up spending a lot of time barking down the wrong rabbit hole due to these changes. @rnk, please see https://github.com/llvm/llvm-project/issues/140799.

rnk commented:
> Would it make sense to have a custom main for AllClangUnitTests and register all targets, just like clang driver does?

That sounds reasonable.

I toyed with the idea of putting the registration into a static initializer, but it raises the question of where you put it. Putting it into an existing test fixture file would work, but it's not nice. Registering targets in main like every other LLVM tool would be the obviously correct thing to do.

kadircet commented:
Do you or @petrhosek have some free cycles to follow up on that soon?

Because as things stand, unit test execution is currently in a very fragile state. Anyone that adds a unittest to clang currently runs the risk of braking something completely unrelated (I do agree that it's the fault of those tests to be fragile in the first place), resulting in lots of time debugging (I spent a good couple hours here).

If we can't get a fix in soon, can we consider reverting this until someone has the capacity to land this with a custom main for clang unit tests?
